graph/floyd_warshall.cpp

```diff
@@ -0,0 +1,148 @@
+ * The resulting shortest path distances between all pairs of vertices are printed to the standard output.
+ * If a negative weight cycle exists, a message is printed instead of the distances.
+ * 
+ * Example input:
+ * 4 4
+ * 0 1 3
+ * 1 3 4
+ * 3 2 2
+ * 2 0 1
+ * 
+ * Example output:
+ * 0 3 7 7
+ * 7 0 4 4
+ * 3 6 0 6
+ * 1 4 2 0
+ */
+
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <limits>
+
+const int INF = std::numeric_limits<int>::max();
+
+/**
+ * @brief Displays the distance matrix.
+ * 
+ * This function prints the distance matrix. For unreachable vertices, it displays "INF".
+ * 
+ * @param graph The matrix representing the shortest path distances between all pairs of vertices.
+ */
+void display(const std::vector<std::vector<int>>& graph) {
+    int n = graph.size();
+    for (int i = 0; i < n; i++) {
+        for (int j = 0; j < n; j++) {
+            if (graph[i][j] == INF) {
+                std::cout << "INF ";
+            } else {
+                std::cout << graph[i][j] << " ";
+            }
+        }
+        std::cout << std::endl;
+    }
+}
+
+/**
+ * @brief Implements the Floyd-Warshall algorithm to find the shortest paths between all pairs of vertices.
+ * 
+ * This function updates the distance matrix `dist` with the shortest path distances by considering
+ * each vertex as an intermediate point. If a shorter path through an intermediate vertex is found,
+ * it updates the distance.
+ * 
+ * @param graph The input adjacency matrix representing direct distances between vertices.
+ * @param dist The matrix where the shortest path distances between all pairs of vertices will be stored.
+ * @return Returns `false` if a negative weight cycle is detected, otherwise returns `true`.
+ */
+bool floyd_warshall(const std::vector<std::vector<int>>& graph, std::vector<std::vector<int>>& dist) {
+    int n = graph.size();
+    for (int k = 0; k < n; k++) {
+        for (int i = 0; i < n; i++) {
+            for (int j = 0; j < n; j++) {
+                if (dist[i][k] < INF && dist[k][j] < INF) {
+                    dist[i][j] = std::min(dist[i][j], dist[i][k] + dist[k][j]);
+                }
+            }
+        }
+    }
+
+    // Check for negative weight cycles
+    for (int i = 0; i < n; i++) {
+        if (dist[i][i] < 0) {
+            return false; // Negative cycle detected
+        }
+    }
+    return true;
+}
+
+/**
+ * @brief Handles the input and invokes the Floyd-Warshall algorithm.
+ * 
+ * This function reads the number of vertices and edges from standard input, builds the adjacency matrix,
+ * and then calls the `floyd` function to compute the shortest path between all pairs of vertices.
+ * If a negative weight cycle is detected, it prints a message; otherwise, it prints the shortest path matrix.
+ */
+void test() {
+    int N, M;
+    std::cout << "Enter the number of vertices and edges: ";
```

Comment:
tests should have predefined cases not user defined cases. 
add cases for the following
1. negative numbers
2. empty array
3. all numbers being INF